dashboard/src/shared/ResourceRef.ts

```diff
@@ -0,0 +1,44 @@
+import { Kube } from "./Kube";
+import { IResource } from "./types";
+
+// ResourceRef defines a reference to a namespaced Kubernetes API Object and
+// provides helpers to retrieve the resource URL
+class ResourceRef {
+  public apiVersion: string;
+  public kind: string;
+  public name: string;
+  public namespace: string;
+
+  // Creates a new ResourceRef instance from an existing IResource. Provide
+  // defaultNamespace to set if the IResource doesn't specify a namespace.
+  // TODO: add support for cluster-scoped resources, or add a ClusterResourceRef
+  // class.
+  constructor(r: IResource, defaultNamespace: string = "default") {
```

Comment:
Hmm good question, the issue here is that if the IResource defines a namespace in `metadata.namespace`, that will always be used. The `defaultNamespace` is only used if `metadata.namespace` is undefined. If you know that the IResource you're passing to `ResourceRef` contains the namespace, you don't need to provide the `defaultNamespace`, if you did, you'd have to set it to `resource.metadata.namespace` which would be a bit annoying.

I do get that this is a bit confusing though, as if `metadata.namespace` is undefined, you probably will need to provide `defaultNamespace`, it's unlikely you'd want that to be "default".

Another option here is to have the signature be `defaultNamespace?: string`, and then in the code we would need to throw an error if both `metadata.namespace` and `defaultNamespace` are undefined. However, I'm not sure it's worth doing that.

Reply:
I guess it's better then to throw the error if `metadata.namespace` and `defaultNamespace` are unset. It could be quite confusing and difficult to debug if suddenly there are unexpected requests to the `default` namespace.

Reply:
done, and added a test - thanks!